### PR TITLE
Configure SQ issues mode for Gradle in PR builds

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 1]
```diff
@@ -2,10 +2,17 @@
 
 import tl = require('vsts-task-lib/task');
 import {ToolRunner} from 'vsts-task-lib/toolrunner';
+import path = require("path");
```
Comment (bgavrilMS) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Not req, I'll remove it.

File: Tasks/Gradle/gradletask.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ import tl = require('vsts-task-lib/task');
 import fs = require('fs');
 import path = require('path');
 
-// Lowercased file names are to lessen the likelihood of xplat issues
+import sqCommon = require('sonarqube-common/sonarqube-common');
```
Comment (bgavrilMS) [inline Tasks/Gradle/gradletask.ts]: Not required, I'll remove it.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 3]
```diff
@@ -2,10 +2,17 @@
...
+
+// Simple data class for a SonarQube generic endpoint
+export class SonarQubeEndpoint {
```
Comment (bgavrilMS) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Moved from lower down.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 4]
```diff
@@ -39,26 +46,41 @@ export function applySonarQubeAnalysisParams(toolRunner:ToolRunner, sqProjectNam
     return toolRunner;
 }
 
-export class SonarQubeEndpoint {
-    constructor(public Url, public Username, public Password) {
+// Run SQ analysis in issues mode, but only in PR builds
+export function applySonarQubeIssuesModeInPrBuild(toolrunner: ToolRunner) {
```
Comment (bgavrilMS) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: The args are identical for both Maven and Gradle so it makes sense to use a shared location for this logic.

File: Tasks/Gradle/gradletask.ts [hunk 5]
```diff
@@ -21,43 +21,11 @@ if (!cwd) {
 }
 tl.cd(cwd);
 
-var gb = tl.createToolRunner(wrapperScript);
+var gradleTool = tl.createToolRunner(wrapperScript);
```
Comment (bgavrilMS) [inline Tasks/Gradle/gradletask.ts]: This file has no functional changes - I just moved some code around

File: Tests/L0/Gradle/_suite.ts [hunk 6]
```diff
@@ -637,18 +747,13 @@ describe('gradle Suite', function() {
             });
     });
 
- it('Gradle with Cobertura and SonarQube', (done) => {
-        setResponseFile('gradleGood.json');
-    
+    it('Gradle with Cobertura and SonarQube', (done) => {
+        setResponseFile('gradleSonarQube.json');
```
Comment (bgavrilMS) [inline Tests/L0/Gradle/_suite.ts]: I moved all the json responses related to SQ to a separate file.

File: Tests/L0/Gradle/gradleGood.json [hunk 7]
```diff
@@ -30,19 +30,7 @@
...
-        },
-        "gradlew build sonarqube -I /gradle/sonar.gradle -Dsonar.host.url=http://sonarqube/end/point -Dsonar.login=uname -Dsonar.password=pword -Dsonar.jdbc.url=jdbc:test:tcp://localhost:8080/sonar -Dsonar.jdbc.username=testDbUsername -Dsonar.jdbc.password=testDbPassword -Dsonar.projectName=test_sqProjectName -Dsonar.projectKey=test_sqProjectKey -Dsonar.projectVersion=test_sqProjectVersion": {
-            "code": 0,
-            "stdout": "Gradle build and SQ analysis with db details done"
-        },
-       "gradlew build cobertura sonarqube -I /gradle/sonar.gradle -Dsonar.host.url=http://sonarqube/end/point -Dsonar.login=uname -Dsonar.password=pword -Dsonar.projectName=test_sqProjectName -Dsonar.projectKey=test_sqProjectKey -Dsonar.projectVersion=test_sqProjectVersion -Dsonar.cobertura.reportPath=CCReport43F6D5EF/coverage.xml": {
-            "code": 0,
-            "stdout": "Gradle build, Cobertura Code Coverage and SQ analysis done"
-        }
+        }
```
Comment (bgavrilMS) [inline Tests/L0/Gradle/gradleGood.json]: I separated some json responses.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 8]
```diff
@@ -39,26 +46,41 @@ export function applySonarQubeAnalysisParams(toolRunner:ToolRunner, sqProjectNam
...
 }
 
 // Gets SonarQube connection endpoint details.
-export function getSonarQubeEndpointFromInput(inputFieldName):SonarQubeEndpoint {
+export function getSonarQubeEndpointFromInput(inputFieldName): SonarQubeEndpoint {
     var errorMessage = "Could not decode the generic endpoint. Please ensure you are running the latest agent (min version 0.3.2)";
     if (!tl.getEndpointUrl) {
-        throw new Error(errorMessage);
+        tl.error(errorMessage);
+        tl.exit(1);
```
Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: I would be careful of this, I know that in at least one location the Maven task tries to hold off on the tl.exit(1) call until later so that code coverage results can be uploaded.

File: Tasks/Gradle/task.json [hunk 9]
```diff
@@ -272,21 +272,14 @@
...
     "instanceNameFormat": "gradlew $(tasks)",
     "execution": {
-        "PowerShell": {
-            "target": "$(currentDirectory)\\gradle.ps1",
-            "argumentFormat": "",
-            "workingDirectory": "$(currentDirectory)",
-            "platforms": [
-                "windows"
-            ]
-        },
```
Comment (HSAR) [inline Tasks/Gradle/task.json]: Is this change intentional?

File: Tests/L0/Gradle/_suite.ts [hunk 10]
```diff
@@ -13,36 +13,62 @@ function setResponseFile(name: string) {
...
+    }
+    else {
+        tr.setInput('sqAnalysisEnabled', 'false');
+    }
+
     return tr;
 }
 
-describe('gradle Suite', function() {
+function setResponseAndBuildVars(initialResponseFile:string, finalResponseFile:string, envVars: Array<[string,string]> ) {
```
Comment (HSAR) [inline Tests/L0/Gradle/_suite.ts]: Is there a particular reason why this method is used rather than keeping it static in the response.json file?

File: Tests/L0/Gradle/_suite.ts [hunk 11]
```diff
@@ -612,12 +638,96 @@ describe('gradle Suite', function() {
             });
     })
 
-    it('Gradle with SonarQube - Should run Gradle with all default inputs when SonarQube analysis disabled', function(done) {
+    it('Gradle with SQ in a PR build - SQ issues mode analysis', function (done) {
+
         // Arrange
-        setResponseFile('gradleGood.json');
+        setResponseAndBuildVars(
+            'gradleSonarQube.json',
```
Comment (HSAR) [inline Tests/L0/Gradle/_suite.ts]: We seem to always use gradleSonarQube.json - is there a risk of dirty reads because of previous tests?

(For the dynamic stuff in the Maven tests I pre-pended this.test.title)

File: Tasks/Gradle/task.json (hunk 9 above)
Comment (bgavrilMS) [inline Tasks/Gradle/task.json]: No, I will revert this!

File: Tests/L0/Gradle/_suite.ts (hunk 10 above)
Comment (bgavrilMS) [inline Tests/L0/Gradle/_suite.ts]: Yes, I think those JSON files make the test very difficult to read, especially as you cannot add comments in Json. This way the intent of the tests is much clearer.
